tests/unittests/Foundation/NSCalendarTests.mm

```diff
@@ -481,15 +472,75 @@
+
+        // January 1st 2010
+        NSDate* constantDate = [calendar dateWithEra:1 year:2010 month:1 day:1 hour:0 minute:0 second:0 nanosecond:0];
+        // January 2nd 2010
+        NSDate* expectedNextDay = [calendar dateWithEra:1 year:2010 month:1 day:2 hour:0 minute:0 second:0 nanosecond:0];
+
+        NSDateComponents* comps = [[NSDateComponents alloc] init];
+        comps.day = 2;
+
+        // No matching options
+        NSDate* date = [calendar nextDateAfterDate:constantDate matchingComponents:comps options:0];
+
+        ASSERT_TRUE_MSG(false, "Should not have gotten here. nextDateAfterDate should throw an INVALID_ARG exception");
+    } @catch (NSException* e) {
+        ASSERT_OBJCEQ(e.name, @"NSInvalidArgumentException");
+    }
+}
+
+OSX_DISABLED_TEST(NSCalendar, NanoSecondComparison) {
+    // Nanosecond comparison is disabled on OSX as the behavior does not match the documentation. Nanosecond comparison will
+    // always return NSOrderedSame if the two dates are within the same second of each other. Thus, even though the two
+    // nanosecond values below are different, OSX will return that this comparison to the granularity of Nanoseconds will
+    // is the same.
+
+    NSCalendar* calendar = [NSCalendar calendarWithIdentifier:NSCalendarIdentifierGregorian];
+    NSDate* date1 = [NSDate dateWithTimeIntervalSinceReferenceDate:.1];
+    NSDate* date2 = [NSDate dateWithTimeIntervalSinceReferenceDate:.9];
+    ASSERT_EQ([calendar compareDate:date1 toDate:date2 toUnitGranularity:NSCalendarUnitNanosecond], NSOrderedAscending);
+    ASSERT_EQ([calendar compareDate:date2 toDate:date1 toUnitGranularity:NSCalendarUnitNanosecond], NSOrderedDescending);
+}
+
+OSX_DISABLED_TEST(NSCalendar, RangeOfWeekOfMonth) {
+    // Retrieving the range of a WeekOfMonth unit should return the number of weeks within the month for that date.
+    // The behavior on OSX however, is that the returned range is always 6 weeks regardless of the actual number of
+    // weeks in the given month.
+
+    NSCalendar* calendar = [NSCalendar calendarWithIdentifier:NSCalendarIdentifierGregorian];
+    NSDate* date = [calendar dateWithEra:1 year:2010 month:3 day:31 hour:3 minute:4 second:5 nanosecond:0];
+    NSRange range = [calendar rangeOfUnit:NSCalendarUnitWeekOfMonth inUnit:NSCalendarUnitYear forDate:date];
+    ASSERT_EQ(range.location, 1);
+    ASSERT_EQ(range.length, 5);
+}
+
+OSX_DISABLED_TEST(NSCalendar, NSCalendarMatchNextTimePreservingSmallerUnitsLeapDay) {
+    // NSCalendarMatchNextTimePreservingSmallerUnits as a matching option does not reflect the documentation accurately.
+    // When matching a date, this option should preserve all smaller values when finding the next date. The behavior on OSX
+    // has instead stripped the date of these smaller values. The date matched from OSX is 2/29/2012 00:00:00 when it should be
+    // 2/29/2012 03:04:05. Thus, this test is disabled.
```

Comment:
This behavior on OSX is not limited to leap days. It always trims the smaller units when it finds an exact match. The documentation is a little ambiguous, but I think this is the intended behavior, and we should probably match it.